### PR TITLE
Variables: Fixes issue with data source variables not updating queries with variable

### LINKED ISSUE #49456: Datasource Selection is broken in recent releases
**What happened**:
I have different prometheus Data sources, all hooked up to a single Grafana instance. This is to give all the users a uniform experience when looking at different panels. In grafana, I have created a template variable (of type datasource) to select between the different datasources.

**What you expected to happen**:
I can select the different data sources from the dropdown but i was expecting the panels will show data for the data source selected. Instead, the panels were blank where i know for sure there should be some data. Also, it was working (8.4.1-1 works fine) and recently got broken (8.5.2-1 and 8.5.3.-1 doesnt work).

I tried to see whats going wrong and here is what i think is the problem.

Let's assume we have datasource variable called **server** that has the values X, Y, Z. This is at the dashboard level. Then we have a datasource selection at the panel level and a datasource shown in each query as well. 
  
In the panel, lets say we have one query A and on the side to it, it shows $server as on the datasource on the panel is selected to be $server. At this point, things work fine. When i select a different data source my query o/p changes, the query A still shows $server on the side. But if i select a specific data source on the panel, then the datasource shown on the query A also changes which is right but now if i set back the datasource to $server, the query A data source is shown the one which is selected in the $server variable. If i select the a different data source on the dashboard, it doesnt refresh the query datasource and the only way to refresh is to select the $server variable  in the panel datasource even though it is already set to $server variable.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Grafana version: 8.5.3-1
- Data source type & version: Prometheus
- OS Grafana is installed on: Centos
- User OS & Browser: MacOS and Chrome
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by torkelo**
Fixes #49456

the change in https://github.com/grafana/grafana/pull/47435 missed that the old solution the DataSourceSettings.uid could be a variable expression , but when we get a DataSourceAPI that is not the case. Very easy mistake to make as this bit is very complex and unintutive.

Related: https://github.com/grafana/support-escalations/issues/2688

### CODE REVIEW

**Review by dprokop — APPROVED** [review]
Sorry about that! 

The fix looks good to me 🔥

**Comment by bobrik** [comment]
I applied this fix, but I'm still seeing this error:

> json decode failure: json: cannot unmarshal object into Go struct field Annotation.annotation.datasource of type string

The query looks like this:

image

**Comment by bobrik** [comment]
After taking a closer look, our error comes from deserializing `datasource` in our internal datasource. I removed it and it works fine, but I'm also not sure whether this is a breaking change for other external datasources.

In simple json datasource plugin it's still just a string:

* https://github.com/grafana/simple-json-datasource#annotation-api

```json
{
  "range": {
    "from": "2016-04-15T13:44:39.070Z",
    "to": "2016-04-15T14:44:39.070Z"
  },
  "rangeRaw": {
    "from": "now-1h",
    "to": "now"
  },
  "annotation": {
    "name": "deploy",
    "datasource": "Simple JSON Datasource",
    "iconColor": "rgba(255, 96, 96, 1)",
    "enable": true,
    "query": "#deploy"
  }
}
```

**Comment by eaufavor** [comment]
We tried this patch and it seems that this fix doesn't prevent my working dashboard from getting into the broken state.

The only way I know to fix it for good is to do the following panel by panel:
https://github.com/grafana/grafana/issues/49456#issuecomment-1136916005.

An ideal fix would prevent me from running into this issue.

**Comment by drstevens** [comment]
@torkelo Thanks for all the hard work maintaining Grafana. It's a great tool.

On Grafana v8.5.6 62355c, we are running into the issue exactly as @eaufavor pointed out. On save any save of a graph in a dashboard using the older `$datasource` syntax, The datasource ID for whatever datasource is selected is silently saved into the individual metrics for the graph. This results in user confusion because the dashboards are showing data for a different datasource than users expect. It would be better to have some failure at least. We spent an embarrassingly long time yesterday wondering why metrics were not as expected because of this.

**Comment by drstevens** [comment]
Are there additional commits in this PR which aren't part of v8.5.6 62355c? I'm wondering if that's the issue. I had been assuming they all were, but it's a bit hard to follow all the cherry-picks
